# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/CanvasRenderingContext2D.cpp
# Commit: 8a94c155e35c
# Full Hash: 8a94c155e35c69a524e0ecb17674bb5afa9c5ad3
# Author: Robert Longson <longsonr@gmail.com>
# Date: 2023-09-12 16:30:15
# Regressor Bug: 1287054
# File Overlap Count: 1
# Description:
#   Bug 1287054 part 1 - support rem, ch, ic and cap units for non-CSS lengths r=emilio,jgilbert
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D180618
# ==============================================================================

diff -r cd564dee524a -r 8a94c155e35c dom/canvas/CanvasRenderingContext2D.cpp
--- a/dom/canvas/CanvasRenderingContext2D.cpp	Mon Sep 11 23:42:37 2023 +0000
+++ b/dom/canvas/CanvasRenderingContext2D.cpp	Mon Sep 11 23:50:11 2023 +0000
@@ -13,6 +13,7 @@
 #include "nsContentUtils.h"
 
 #include "mozilla/intl/BidiEmbeddingLevel.h"
+#include "mozilla/GeckoBindings.h"
 #include "mozilla/PresShell.h"
 #include "mozilla/PresShellInlines.h"
 #include "mozilla/SVGImageContext.h"
@@ -2798,37 +2799,63 @@
 class CanvasUserSpaceMetrics : public UserSpaceMetricsWithSize {
  public:
   CanvasUserSpaceMetrics(const gfx::IntSize& aSize, const nsFont& aFont,
-                         nsAtom* aFontLanguage, bool aExplicitLanguage,
+                         const ComputedStyle* aCanvasStyle,
                          nsPresContext* aPresContext)
       : mSize(aSize),
         mFont(aFont),
-        mFontLanguage(aFontLanguage),
-        mExplicitLanguage(aExplicitLanguage),
+        mCanvasStyle(aCanvasStyle),
         mPresContext(aPresContext) {}
 
-  virtual float GetEmLength() const override {
-    return mFont.size.ToCSSPixels();
-  }
-
-  virtual float GetExLength() const override {
-    nsFontMetrics::Params params;
-    params.language = mFontLanguage;
-    params.explicitLanguage = mExplicitLanguage;
-    params.textPerf = mPresContext->GetTextPerfMetrics();
-    params.featureValueLookup = mPresContext->GetFontFeatureValuesLookup();
-    RefPtr<nsFontMetrics> fontMetrics =
-        mPresContext->GetMetricsFor(mFont, params);
-    return NSAppUnitsToFloatPixels(fontMetrics->XHeight(),
-                                   AppUnitsPerCSSPixel());
-  }
-
-  virtual gfx::Size GetSize() const override { return Size(mSize); }
+  float GetEmLength(Type aType) const override {
+    switch (aType) {
+      case Type::This:
+        return mFont.size.ToCSSPixels();
+      case Type::Root:
+        return SVGContentUtils::GetFontSize(
+            mPresContext->Document()->GetRootElement());
+      default:
+        MOZ_ASSERT_UNREACHABLE("Was a new value added to the enumeration?");
+        return 1.0f;
+    }
+  }
+  gfx::Size GetSize() const override { return Size(mSize); }
 
  private:
+  GeckoFontMetrics GetFontMetricsForType(Type aType) const override {
+    switch (aType) {
+      case Type::This: {
+        if (!mCanvasStyle) {
+          return DefaultFontMetrics();
+        }
+        return Gecko_GetFontMetrics(
+            mPresContext, WritingMode(mCanvasStyle).IsVertical(),
+            mCanvasStyle->StyleFont(), mCanvasStyle->StyleFont()->mFont.size,
+            /* aUseUserFontSet = */ true,
+            /* aRetrieveMathScales */ false);
+      }
+      case Type::Root:
+        return GetFontMetrics(mPresContext->Document()->GetRootElement());
+      default:
+        MOZ_ASSERT_UNREACHABLE("Was a new value added to the enumeration?");
+        return DefaultFontMetrics();
+    }
+  }
+
+  WritingMode GetWritingModeForType(Type aType) const override {
+    switch (aType) {
+      case Type::This:
+        return WritingMode(mCanvasStyle);
+      case Type::Root:
+        return GetWritingMode(mPresContext->Document()->GetRootElement());
+      default:
+        MOZ_ASSERT_UNREACHABLE("Was a new value added to the enumeration?");
+        return WritingMode();
+    }
+  }
+
   gfx::IntSize mSize;
   const nsFont& mFont;
-  nsAtom* mFontLanguage;
-  bool mExplicitLanguage;
+  RefPtr<const ComputedStyle> mCanvasStyle;
   nsPresContext* mPresContext;
 };
 
@@ -2858,12 +2885,18 @@
     return;
   }
 
+  RefPtr<const ComputedStyle> canvasStyle;
+
   // The PresContext is only used with URL filters and we don't allow those to
   // be used on worker threads.
   nsPresContext* presContext = nullptr;
   if (presShell) {
     if (FiltersNeedFrameFlush(CurrentState().filterChain.AsSpan())) {
       presShell->FlushPendingNotifications(FlushType::Frames);
+      if (mCanvasElement) {
+        canvasStyle =
+            nsComputedDOMStyle::GetComputedStyleNoFlush(mCanvasElement);
+      }
     }
 
     if (MOZ_UNLIKELY(presShell->IsDestroying())) {
@@ -2877,9 +2910,8 @@
 
   CurrentState().filter = FilterInstance::GetFilterDescription(
       mCanvasElement, CurrentState().filterChain.AsSpan(), writeOnly,
-      CanvasUserSpaceMetrics(GetSize(), CurrentState().fontFont,
-                             CurrentState().fontLanguage,
-                             CurrentState().fontExplicitLanguage, presContext),
+      CanvasUserSpaceMetrics(GetSize(), CurrentState().fontFont, canvasStyle,
+                             presContext),
       gfxRect(0, 0, mWidth, mHeight), CurrentState().filterAdditionalImages);
   CurrentState().filterSourceGraphicTainted = writeOnly;
 }